### DISCUSSION TITLE
Integration with existing trace libraries

### DISCUSSION THREAD

**Discussion by adam-mccormick** [moleculerjs/moleculer#1062] [Ideas]
Projects like [opentelmetry](https://opentelemetry.io/) provide prebuilt instrumentation adapters for a whole range of libraries like MySQL, express, cloud SDKs, socket.io etc.  It would convenient to use these tools in Moleculer without having to abandon the built-in tracing API provided by the framework.

Perhaps there's a way to share the current trace and span context between Moleculer and another library like OpenTelemetry and allow both to update the trace context. 

A better idea might be to have a configurable `SpanFactory` that can create span instances that follow the existing Moleculer API but can use delegate span objects from other frameworks.

**Comment by icebob** [comment]
The Moleculer built-in tracing feature is a complete solution like the OpenTelemetry lib. Yeah, Moleculer Tracer doesn't have instrumentation for external libs and I don't plan to add it because it's a big "hacking" job, a lot of time, and doesn't give extra to the framework itself. 

Instead, my suggestion is to create a Moleculer instrumentation for OpenTelemetry. Similar to what [dd-trace](https://github.com/DataDog/dd-trace-js/pull/1598) does.

It can be a middleware ([based on the built-in tracing middleware](https://github.com/moleculerjs/moleculer/blob/master/src/middlewares/tracing.js)) which wraps the `localAction` and `localEvent` hooks and create OT spans instead of Moleculer Tracer spans.

**Reply by intech** [reply]
I agree that this is a lot of work.

I have an idea of how we can solve this by creating a bridge for connecting external shimmer/wrap tracing modules, according to the principle, as it is now done in the [Elastic APM node.js agent](https://github.com/elastic/apm-agent-nodejs/tree/main/lib/instrumentation).
The idea is to forward moleculer tracing methods and they can interact with these modules through a normal connection.

I’m also waiting and hoping for the [node.js API](https://nodejs.org/api/tracing.html) that in the end, everything will come to a single channel for application tracing, and we won’t have to invent our own solution in each library.

**Comment by icebob** [comment]
I made an OpenTelemetry middleware and a demo Moleculer project which demonstrates my suggestion. It uses B3 propagator which pushes tracing info into `ctx.meta`, so it can connect the traces of remote actions as well. And also can connect the trace spans between built-in instruments and Moleculer action calls.

https://github.com/icebob/moleculer-opentelemetry-demo

![image](https://user-images.githubusercontent.com/306521/154852289-c2f4e052-67f3-46dc-9740-20184a82f191.png)

The main things are in:
- https://github.com/icebob/moleculer-opentelemetry-demo/blob/master/opentelemetry.middleware.js
- https://github.com/icebob/moleculer-opentelemetry-demo/blob/master/moleculer.config.js

_My experiences with OT:_
1. OpenTelemetry documentation is very simple, not easy to make complex solutions. I had to check the source code many times to find out how it works and it took ~6 hours from my time to make a working demo project :(
2. Jaeger accepts the service name only once (in `getSender` which generates only one per process), so this is why all spans' service is `moleculer`. Maybe Zipkin exporter is better.

**Reply by vkruoso** [reply]
Hi @icebob @intech, thanks a lot for the work. Was looking around your demo and trying to figure out a way to get the span service name to be the correct one. Trying with Zipkin exporter and Jaeger, gives the result below, which looks better. It is just missing the correct name for the specific OT libraries spans. Maybe is related to this: https://github.com/open-telemetry/opentelemetry-js/issues/2606. If you have any pointers, would be appreciated. If I manage to get it working will let you know!

![Screen Shot 2022-03-04 at 00 12 47](https://user-images.githubusercontent.com/293572/156692620-bd0764a0-8a1d-4f77-a314-8c8912141db5.png)

**Reply by icebob** [reply]
Yeah, this is the reason. Jaeger exporter creates a `Sender` which contains the service name and this sender cached and not created again. So all spans from the given process will use the same service name. This issue is solved in the built-in Moleculer Jaeger exporter here: https://github.com/moleculerjs/moleculer/blob/1fa2440279ec8f242399e1f3dea30abead7c66e0/src/tracing/exporters/jaeger.js#L162-L177

**Reply by vkruoso** [reply]
Sorry for the delay here. Thanks a lot @icebob and others here. I did what you suggested and it worked for us! We are running only one service per process at this moment, so it simplifies things a little, but tracing is working nicely with NewRelic, and locally with Jaeger as well using the Zipkin format. As a side note, just had to do small changes to your code in the error handling part (to close the span to correctly report the error) and in some parameters of the traces as NewRelic was not linking some of them.

**Reply by daniel-white** [reply]
Hey @icebob would you consider making this part of OT's contrib repo? That way it can grow with the other instrumentation packages

**Comment by Aaron-Ritter** [comment]
Nice example! btw. you could use the standard OTLP exporter and run e.g. https://raw.githubusercontent.com/grafana/docker-otel-lgtm/refs/heads/main/run-lgtm.sh locally. Same topic with 1 process 1 service but one less component in the mix.

by the way one of the reasons why you see a single span rather than multiple ones is described well here https://stackoverflow.com/questions/72867313/opentelemetry-open-span-in-one-process-and-close-it-in-another. and this is partially just because how moleculer works. And at some point profiling will allow more detailed insight into the different spans https://opentelemetry.io/blog/2024/profiling/
